gsy_framework/sim_results/carbon_emissions.py

```diff
@@ -0,0 +1,246 @@
+            raise ValueError("stat must be one of 'min', 'median', or 'max'")
+
+        df = self._query_generation_per_plant(country_code, start, end)
+
+        generation_types = df.iloc[0].iloc[1:]
+        carbon_emission_per_plant = lambda x: GENERATION_PLANT_TO_CARBON_EMISSIONS[x][
+            stat
+        ]  # noqa: E731 C3001
+        emissions_map = generation_types.map(carbon_emission_per_plant)
+
+        df_numeric = df.iloc[2:].reset_index(drop=True)
+        df_numeric = df_numeric.set_index("Unnamed: 0").apply(  # noqa: E501
+            pd.to_numeric, errors="coerce"
+        )
+        df_numeric.index = pd.to_datetime(df_numeric.index)
+        df_numeric["Total Energy (kWh)"] = df_numeric.sum(axis=1) * 1000  # convert MWh to kWh
+        df_numeric["Total Carbon (gCO2eq)"] = df_numeric.mul(emissions_map, axis=1).sum(axis=1)
+        df_numeric["Ratio (gCO2eq/kWh)"] = (
+            df_numeric["Total Carbon (gCO2eq)"] / df_numeric["Total Energy (kWh)"]
+        )
+        return df_numeric[["Ratio (gCO2eq/kWh)"]]
+
+    def calculate_carbon_emissions_from_imported_energy(
+        self, df_carbon_ratio: pd.DataFrame, df_imported_energy: pd.DataFrame
+    ):
+        """Calculate the carbon emissions based on the imported/exported energy"""
+
+        df_carbon_ratio.index = pd.to_datetime(df_carbon_ratio.index)
+        df_imported_energy["simulation_time"] = pd.to_datetime(
+            df_imported_energy["simulation_time"]
+        )
+        df_combined = pd.merge(
+            df_imported_energy,
+            df_carbon_ratio,
+            left_on="simulation_time",
+            right_index=True,
+            how="left",
+        )
+        df_combined["carbon_generated_base_case"] = (
+            df_combined["imported_from_community"] + df_combined["imported_from_grid"]
+        ) * df_combined["Ratio (gCO2eq/kWh)"]
+        df_combined["carbon_generated_gsy"] = (
+            df_combined["imported_from_grid"] * df_combined["Ratio (gCO2eq/kWh)"]
+        )
+        df_combined["carbon_savings"] = (
+            df_combined["carbon_generated_base_case"] - df_combined["carbon_generated_gsy"]
+        )
+        df_combined = df_combined[
+            [
+                "simulation_time",
+                "imported_from_community",
+                "imported_from_grid",
+                "carbon_generated_base_case",
+                "carbon_generated_gsy",
+                "carbon_savings",
+            ]
+        ]
+        return df_combined
+
+    def update(self, country_code: str, time_slot: duration, results_list: List[Dict]):
+        """Update the results_dict with carbon generated and carbon savings"""
+
+        start_date = pd.Timestamp(results_list[0]["current_market"].split("T")[0])
+        end_date = pd.Timestamp(results_list[-1]["current_market"].split("T")[0])
+        df_carbon_ratio = self.calculate_carbon_ratio(country_code, start_date, end_date)
+        df_carbon_ratio.to_csv("carbon_ratio.csv")
```

Comment:
I assume that this is for debugging right ? Please remember to remove before merging

Reply:
Yes, removed, thanks